app/src/main/java/org/mozilla/fenix/perf/NavGraphProvider.kt

```diff
@@ -30,15 +45,42 @@ object NavGraphProvider {
     // We want to store member state on the NavController. However, there is no way to do this.
     // Instead, we store state as part of a map: NavController instance -> State. In order to
     // garbage collect our data when the NavController is no longer relevant, we use a WeakHashMap.
-    private val map = WeakHashMap<NavController, Job>()
+    private val map = WeakHashMap<NavController, DeferredNavGraphContainer>()
 
-    fun inflateNavGraphAsync(navController: NavController, lifecycleScope: LifecycleCoroutineScope) {
-        val inflationJob = lifecycleScope.launch(Dispatchers.IO) {
-            val inflater = navController.navInflater
-            navController.graph = inflater.inflate(R.navigation.nav_graph)
+    private data class DeferredNavGraphContainer(
+        val inflationJob: Deferred<NavGraph>?,
+        var isGraphSet: Boolean = false
+    )
+
+    @OptIn(ExperimentalCoroutinesApi::class) // UNDISPATCHED.
+    fun inflateNavGraphAsync(
+        navController: NavController,
+        lifecycleScope: CoroutineScope,
+        ioDispatcher: CoroutineDispatcher = Dispatchers.IO
```

Comment:
@MarcLeclair AIUI, inflation reads the layout file from disk and then heavily uses CPU to complete the process. Going from this simple point of view, I think the IO dispatcher is more fitting.

From a more nuanced perspective, we should be careful about using the `Default` dispatcher for urgent operations because if the `Default` dispatcher gets too many jobs, newer jobs are blocked. I'd consider inflating the core UI very urgent so I'm hesitant to use it. I think if the Default dispatcher was implemented with more nuance, it might be an option. Specifically, if you're going to have a limited number of simultaneous jobs being executed (which is optimal for CPU-bound work), the jobs should be able to specify their relative priorities. For example, inflating the UI might be P1 and it'd bump any active P2 jobs from the queue until it completed. But that's not how the Default dispatcher is implemented so it's of limited utility.

What do you think?

Reply:
Spoke with MarcLeclair: we're okay with the IO dispatcher. MarcLeclair raised a point that we should try to set a good example by not defaulting to the IO dispatcher but we agreed unless we came up with a standard threading model for the app, it would be difficult to not make the IO dipatcher the default given the lack of a priority queue in the Default dispatcher.

tldr; we're fine with this change.